### PR TITLE
Fix for "268 - Clear search from details page with trash icon"

### PR DESCRIPTION
**Pull request by kghmanuel**
This is a minimalist set of changes that would address the issues reported. An alternative is to start injecting and using MLRemoteInputService and SearchModel at different areas.

### CODE REVIEW

File: ui/app/search/search-ctrl.js [hunk 1]
```diff
@@ -26,30 +26,22 @@
         var oldParams = mlSearch.getParams();
         var newParams = mlSearch.getCurrentParams();
 
-        // restore params, if we have cached results, and no params provided
-        if (model.search.results && _.isEqual( {}, newParams )) {
+        // capture URL params in mlSearch and ctrl model
+        mlSearch.fromParams().then(function() {
 
-          $location.search( oldParams );
```
Comment (kghmanuel) [inline ui/app/search/search-ctrl.js]: note that this will actually trigger '$locationChangeSuccess', thus triggering a search nevertheless.

File: ui/app/search/search-ctrl.js [hunk 2]
```diff
@@ -26,30 +26,22 @@
...
-
-        } else {
-
-          // capture URL params in mlSearch and ctrl model
```
Comment (kghmanuel) [inline ui/app/search/search-ctrl.js]: these lines weren't really removed, just re-indented.

File: ui/app/search/search-ctrl.js (hunk 1 above)
Comment (grtjn) [inline ui/app/search/search-ctrl.js]: I think that was the intention, to trigger locationChangeSuccess. That has an extra check to only do an actual search if url changed. You think that check was flawed?

File: ui/app/search/search-ctrl.js [hunk 3]
```diff
@@ -87,10 +79,11 @@
         mlSearch.addAdditionalQuery(combinedQuery);
         search();
       }
-
+      
       function search(qtext) {
         if ( arguments.length ) {
           model.qtext = qtext;
+          $location.search().q = qtext;
```
Comment (grtjn) [inline ui/app/search/search-ctrl.js]: Why is this necessary?

File: ui/app/search/search-ctrl.js [hunk 4]
```diff
@@ -26,30 +26,22 @@
...
-        }
+        });
 
         // capture URL params (forward/back, etc.)
         $scope.$on('$locationChangeSuccess', function(e, newUrl, oldUrl){
           if (newUrl !== oldUrl) {
             mlSearch.locationChange( newUrl, oldUrl ).then(function() {
-              search();
+              //TODO: how do we reflect it back to the directive?
+              search(model.mlSearch.getCurrentParams().q);
```
Comment (grtjn) [inline ui/app/search/search-ctrl.js]: Why is this necessary? (just trying to understand..)

File: ui/app/search/search-ctrl.js (hunk 3 above)
Comment (kghmanuel) [inline ui/app/search/search-ctrl.js]: The url that may contain a q will force the search later on to make use of that old value. The alternative to this is to modify the directive's controller or override the directive's behavior.

However, that change would require various wiring that imo, introduces too much complexity relative to the goal.

File: ui/app/search/search-ctrl.js (hunk 4 above)
Comment (kghmanuel) [inline ui/app/search/search-ctrl.js]: The model is no longer in synch with the url. If i am not mistaken, the alternative is to inject remoteInput, SearchModel, etc and synch them all beforehand.

File: ui/app/search/search-ctrl.js (hunk 1 above)
Comment (kghmanuel) [inline ui/app/search/search-ctrl.js]: coming from demo details page, the "show all demos" will shift from details to an empty search (i.e. no ?q=oldsearch). given the condition above:

```
_.isEqual( {}, newParams )
```

this will execute and force the url to change. in effect, 2 searches were performed.

Comment (grtjn) [inline ui/app/search/search-ctrl.js]: Perhaps easiest to share screen, and discuss together. Do you have time tomorrow?

**Comment by kghmanuel** [comment]
much more broken than expected. will investigate further.
